### PR TITLE
CON-497: Implement Core Functionality

### PR DESCRIPTION
**Pull request by CarstenHess**
The service now does the following:

- Fetch all configurations, "oldest last-run first" (with the brand new ones that have never been handled on top)
  - For each configuration, look in the analytics "Link" table to get all matching urls that are newer than "last run" (or 3 month back if never run before), and removes *all* query parameters.
    - For each url, check if it's internal or already has been crawlend. If not, order a page check.
  - Updates the configuations "last run" with the current timestamp and "last page found" if any page crawls were requested.

### CODE REVIEW

File: Alpine.Service/Service/LandingPageFinder.cs [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+							.Param("timestamp", timestamp)
+							.Param("integration_id", configId)
+						).ExecuteAsync();
+					}
+				}
+			}
+			catch (SqlException ex)
+			{
+				log.Error(ex, "SqlException while updating timestamps for config id: {ConfigId}, with results: {WithResults}", configId, withResults);
+				throw;
```
Comment (ricflams) [inline Alpine.Service/Service/LandingPageFinder.cs]: This will abort the full landingpage-integrations-loop.

Maybe we should handle each landingpage-integration as one "isolated" case? Meaning here, if one of them fails then it wouls not impact the others. There could perhaps be just one "bad apple" for some reason.

For instance by extracting the inner code of the `foreach (var integration in integrations)` as a method, whose body was in a try-catch but which would _not_ throw an exception upon error.

I would propose extracting it into a method simply because the number of try-catches and loops in `Run()` would likely become a bit too unwieldly by adding one more try-catch: two try-catch'es with two loops is a bit hard to follow.

File: Alpine.Service/Service/LandingPageFinder.cs [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+			{
+				// If this becomes too much for each service "loop" - it is safe to
+				// limit this select (to get only the first n configs) since it takes
+				// the "oldest" first and updates timestamp when they are handled
+				using (var conn = _sqlClient.Connect())
+				{
+					return await conn.Do(s => s.Raw(@"
+					select
+						id Id
+						,account_id AccountId
```
Comment (ricflams) [inline Alpine.Service/Service/LandingPageFinder.cs]: Oh, my eyes. Not a fan of this notation.
Yes, I can see that's it's easy to add one more line at the end with just a one-line change. But no, it simply looks ugly, regardless of whether it's SQL-style or not.

File: Alpine.Service/Service/LandingPageFinder.cs [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+			}
+		}
+
+		private async Task<bool> OrderPageCheck(long accountId, string url, long siteId, CancellationToken ct)
+		{
+			var log = _logZone.ForMethod();
+
+			if (string.IsNullOrWhiteSpace(url))
+			{
+				log.Warn("Url was null or empty - AccountId: {AccountId}, SiteId: {SiteId}", accountId);
```
Comment (ricflams) [inline Alpine.Service/Service/LandingPageFinder.cs]: You're missing the `siteId` param.

File: Alpine.Service/Service/LandingPageFinder.cs [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+								union all
+									(select id from [schema].page_[site_id] where lower(url) in(:url, :trimmed))")
+							.Param("site_id", siteId)
+							.Param("url", lurl)
+							.Param("trimmed", lurl.TrimEnd('/')))
+						.SingleAsync<long?>().ConfigureAwait(false);
+
+					if (id.HasValue)
+					{
+						log.Verbose("Found existing page: {Url} - AccountId: {AccountId}, SiteId: {SiteId}", url, accountId, siteId);
```
Comment (ricflams) [inline Alpine.Service/Service/LandingPageFinder.cs]: Here too, `.Debug()` might be a better level.

File: Alpine.Service/Service/LandingPageFinder.cs [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+						return false;
+					}
+
+					var result = await conn.Do(x => x
+							.Raw("select [schema].lcdomain.is_url_internal(:url, :site_id) test from dual")
```
Comment (ricflams) [inline Alpine.Service/Service/LandingPageFinder.cs]: `is_internal` or something might be a better name than simply `test`

File: Alpine.Service/Models/LandingPageIntegration.cs [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+	{
+		public long Id { get; set; }
+		public long AccountId { get; set; }
+		public LandingPageIntegrationType IntegrationType { get; set; }
+		public long? TaggingPolicyId { get; set; }
+		public string UrlPattern { get; set; }
+		public long AnalyticsSiteId { get; set; }
+		public long QaSiteId { get; set; }
+		public DateTime? LastRunTime { get; set; }
+		public DateTime? LastResultFoundTime { get; set; }
```
[the remark this answers is not in the document]
Comment (CarstenHess) [reply Alpine.Service/Models/LandingPageIntegration.cs]: I really think this would be overkill to spend time on -- it is only useful in very specific situations and for users who don't know their own Hubspot/whatever setup.

This is also  information the users should go check in Aanalytics if the want to know about it -- plus we should definately not expose analytics-data to customers that only have priority, so we would have to make that column would only available to analytics customers... I think this is better fixed by mentioning where the info is in  Analytics, on the support-pages for this functionality.

File: Alpine.Service/Service/LandingPageFinder.cs [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+using Alpine.Service.Utils.Options;
+using Alpine.Service.Utils.Oracle;
+using Sql.Client;
+using SqlSvc;
+
+namespace Alpine.Service.Service
+{
+	public class LandingPageFinder : ILandingPageFinder
+	{
+		private static readonly TimeSpan CheckInterval = TimeSpan.FromSeconds(10);
```
[the remark this answers is not in the document]
Comment (CarstenHess) [reply Alpine.Service/Service/LandingPageFinder.cs]: I agree that the 10s is too low, Cavit said 10m would be ok

File: Alpine.Service/Service/LandingPageFinder.cs [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+		{
+			 _logZone = logFactory.CreateLogZone(this);
+			_healthChecker = healthChecker;
+			_sqlClient = new SqlClient(options.SqlGatewayHost);
+		}
+
+		public async Task Run(CancellationToken ct)
+		{
+			var log = _logZone.ForMethod();
+			var heartbeat = _healthChecker.CreateHeartbeat("LandingPageFinder", CheckInterval);
```
[the remark this answers is not in the document]
Comment (CarstenHess) [reply Alpine.Service/Service/LandingPageFinder.cs]: when the waiting time is 10m instead of 10s it should def. be doable

[the remark this answers is not in the document]
Comment (ricflams) [reply Alpine.Service/Service/LandingPageFinder.cs]: Meaning, doable to call heartbeat inside the loop - or doable to do all the work within the 60s heartbeat-grace-period?

File: Alpine.Service/Models/LandingPageIntegration.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (ricflams) [reply Alpine.Service/Models/LandingPageIntegration.cs]: Fair enough

File: Alpine.Service/Service/LandingPageFinder.cs [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+
+			log.Info("Started");
+			while (!ct.IsCancellationRequested)
+			{
+				try
+				{
+					log.Verbose("Checking for landing page configs");
```
[the remark this answers is not in the document]
Comment (CarstenHess) [reply Alpine.Service/Service/LandingPageFinder.cs]: There is def. something about the log levels here that seems off -- but this (and the `Verbose()` just below) should actually be `Info()` instead, whereas the `Info()`s should be `Debug()`.
I'll also try to set the default log level to Debug when in debug-mode.

File: Alpine.Service/Service/LandingPageFinder.cs [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+					foreach (var integration in integrations)
+					{
+						log.Info("Looking for new pages since {lastChecked} for pattern: {UrlPattern}, AccountId: {AccountId}", integration.LastRunTime, integration.UrlPattern, integration.AccountId);
+						var currentTime = DateTime.Now;
+						var urls = await FindPageMatchUrls(integration.AccountId, integration.AnalyticsSiteId, integration.LastRunTime, integration.UrlPattern);
+						log.Info("Found {urls} number new urls for pattern: {UrlPattern}, AccountId: {AccountId}", urls.Count, integration.UrlPattern, integration.AccountId);
+						var didOrderAPageCheck = false;
+						foreach (var url in urls)
+						{
+							didOrderAPageCheck |= await OrderPageCheck(integration.AccountId, url, integration.QaSiteId, ct);
```
[the remark this answers is not in the document]
Comment (CarstenHess) [reply Alpine.Service/Service/LandingPageFinder.cs]: But for 2 booleans, `|=`  is really not a bitwise operator ... its just a weird syntactical C# quirk that it's not supposed to be written as `||=` -- I hope you don't have anything against using logical operators on booleans in general ;-)  

When you realize that `x |= y` is LITERALLY the same as writing `x = x || y` when x & y are both booleans it really shouldn't bother you.

But maybe it's just a feeling-thing my aversion to the ternary ?: shorthand notation, that I also *know* is valid and correct, but still it annoys me

File: Alpine.Service/Service/LandingPageFinder.cs [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+						).ExecuteAsync();
+					}
+					else
+					{
+						await conn.Do(s => s.Raw(@"
+						update
+							landing_page_integration
+						set
+							last_run_time = TO_DATE(:timestamp,:datemask),
+							last_result_found_time = TO_DATE(:timestamp,:datemask)
```
[the remark this answers is not in the document]
Comment (CarstenHess) [reply Alpine.Service/Service/LandingPageFinder.cs]: Already did that :-)

File: Alpine.Service/Service/LandingPageFinder.cs (hunk 1 above)
Comment (CarstenHess) [reply Alpine.Service/Service/LandingPageFinder.cs]: Very true - it's not intentional to abort the main loop for any other reason than the service being shut down.

I'll do something about this!

File: Alpine.Service/Service/LandingPageFinder.cs (hunk 3 above)
Comment (CarstenHess) [reply Alpine.Service/Service/LandingPageFinder.cs]: yikes!

File: Alpine.Service/Service/LandingPageFinder.cs (hunk 4 above)
Comment (CarstenHess) [reply Alpine.Service/Service/LandingPageFinder.cs]: yup

File: Alpine.Service/Service/LandingPageFinder.cs (hunk 2 above)
Comment (CarstenHess) [reply Alpine.Service/Service/LandingPageFinder.cs]: all I did wass have our *only common sql* tool format this for me ... but I'll change it

File: Alpine.Service/Service/LandingPageFinder.cs (hunk 5 above)
Comment (CarstenHess) [reply Alpine.Service/Service/LandingPageFinder.cs]: I copied this sql directly from pfg - i didn't see this weird naming -- maybe they though that the name is apt because it is a test for existence of the page ?

File: Alpine.Service/Service/LandingPageFinder.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (CarstenHess) [reply Alpine.Service/Service/LandingPageFinder.cs]: both - it makes sense to perform the heatbeat inside, to work around timeouts if we get a lot of configurations, but in general it *should* be feasible to handle everything we can throw against the service within 10m.

File: Alpine.Service/Service/LandingPageFinder.cs (hunk 1 above)
Comment (CarstenHess) [reply Alpine.Service/Service/LandingPageFinder.cs]: I moved stuff around a bit - it should be better now

File: Alpine.Service/Models/LandingPageIntegration.cs [hunk 12]
```diff
@@ -0,0 +1,24 @@
+﻿using System;
+
+namespace Alpine.Service.Models
+{
+	public enum LandingPageIntegrationType
```
[the remark this answers is not in the document]
Comment (CarstenHess) [reply Alpine.Service/Models/LandingPageIntegration.cs]: agree

File: Alpine.Service/Service/LandingPageFinder.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (ricflams) [reply Alpine.Service/Service/LandingPageFinder.cs]: Yes, but you also _wait_ 10 min between in each loop. So the work has to be done within the "grace period" of 60s of the heartbeat if we don't want it to report unhealthy.

I think pinging the heartbeat inside the loop could be a sensible solution.

File: Alpine.Service/Service/LandingPageFinder.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (ricflams) [reply Alpine.Service/Service/LandingPageFinder.cs]: Yes, I know the operator works for bools too. But the single-pipe operator just shout out "bit mask operations!" to me. Maybe irrationally.

File: Alpine.Service/Options/ServiceOptions.cs [hunk 13]
```diff
@@ -9,7 +9,14 @@ public class ServiceOptions : IServiceOptions
 		[Option("service-port", HelpText = "Offer HTTP service access on this port", Default = 8484)]
 		public int ServicePort { get; set; }
 
-		[Option("log-level", HelpText = "Use this default log level", Default = LogLevel.Info)]
+
+		[Option("log-level", HelpText = "Use this default log level",
+#if DEBUG
+			Default = LogLevel.Info
```
Comment (ricflams) [inline Alpine.Service/Options/ServiceOptions.cs]: I reckon you mean `LogLevel.Debug` here? 😉

File: Alpine.Service/Service/LandingPageFinder.cs [hunk 14]
```diff
@@ -42,31 +42,48 @@ public async Task Run(CancellationToken ct)
...
-				heartbeat.Beat();
 				await Task.Delay(CheckInterval, ct);
 			}
 
 			log.Info("Stopped");
 		}
 
+		private async Task HandleSingleLandingPageIntegration(CancellationToken ct, ILogger log, LandingPageIntegration integration)
+		{
+			if (integration == null)
```
Comment (ricflams) [inline Alpine.Service/Service/LandingPageFinder.cs]: Is this even possible? I reckon not.
I'm not a fan of gratuitous null-checks, that sort of suggests that maybe, just maybe, we could expect to see a null here.

File: Alpine.Service/Service/LandingPageFinder.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (CarstenHess) [reply Alpine.Service/Service/LandingPageFinder.cs]: I do admit that the syntax seems weird at best

File: Alpine.Service/Service/LandingPageFinder.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (CarstenHess) [reply Alpine.Service/Service/LandingPageFinder.cs]: I definitly *don't* want to wait 10m between each loop -- I want it to do all configurations in one flow, and then sleep for 10m -- but I want to keep the heartbeat up to date, so it will not get killed if the sum time of handling all exceeds the (not visible) 1m grace period - but if one of the single configs take >11m or we fail previous to doing anything in the loop twice, something is so wrong that the service should be considered sick.

File: Alpine.Service/Service/LandingPageFinder.cs [hunk 15]
```diff
@@ -0,0 +1,308 @@
...
+
+					await conn.Do(s => s.Raw(@"
+						update
+							landing_page_integration
+						set
+							last_run_time = TO_DATE(:timestamp,:datemask)
+							[set_found_time]
+						where
+							id = :integration_id")
+						.Param("timestamp", timestamp)
```
Comment (andreyeurope) [inline Alpine.Service/Service/LandingPageFinder.cs]: Can't we add the `withResults` in an `.If` method together with the content of `setFoundTime` instead of the variable and the above if statement?

Comment (ricflams) [reply Alpine.Service/Service/LandingPageFinder.cs]: It doesn't seem like the `SqlClient` offer `If` etc 😢

Comment (andreyeurope) [reply Alpine.Service/Service/LandingPageFinder.cs]: Damn.. Well, then everything is as good as it gets.

Comment (CarstenHess) [reply Alpine.Service/Service/LandingPageFinder.cs]: I liked that suggestion though - it would have been neat
